src/algebra/order/rearrangement.lean

```diff
@@ -0,0 +1,376 @@
+/-
+Copyright (c) 2021 Mantas Bakšys. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mantas Bakšys
+-/
+import algebra.order.smul
+import group_theory.perm.sign
+import order.monovary
+import tactic.abel
+
+/-!
+# Rearrangement inequality
+
+This file proves the rearrangement inequality.
+
+The rearrangement inequality tells you that for two functions `f g : ι → α`, the sum
+`∑ i, f i * g (σ i)` is maximized over all `σ : perm ι` when `g ∘ σ` monovaries with `f` and
+minimized when `g ∘ σ` antivaries with `f`.
+
+The case for `monotone` pairs of functions over a `linear_order` is deduced.
+
+Note : I'm currently very unsure if this theorem should have it's own file, please feel free to
+comment on this
+-/
+
+variables {ι ι' α β γ : Type*}
+
+section vary
+variables [preorder α] [preorder β] {f : ι → α} {g : ι → β}
+
+open equiv
+
+end vary
+
+namespace finset
+
+/-- Induction principle for `finset` based on the ordering induced by a function. -/
+lemma induction_on_max_value [linear_order α] [decidable_eq ι] (f : ι → α)
+  {p : finset ι → Prop} (s : finset ι) (h0 : p ∅)
+  (step : ∀ a s, a ∉ s → (∀ x ∈ s, f x ≤ f a) → p s → p (insert a s)) : p s :=
+begin
```

Comment:
How does this differ from `finset.induction_on_max`? Can you show one in terms of the other? If not, can you add docstrings saying when one is preferable over the other?

Reply:
`finset.induction_on_max` can  have the assumption that `∀ x ∈ s, x < a` but the analogue is not the case for functions, which are not injective. I'll add this to a docstring